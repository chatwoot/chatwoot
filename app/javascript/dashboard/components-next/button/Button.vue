<script setup>
import { computed, useSlots } from 'vue';

import Spinner from 'dashboard/components-next/spinner/Spinner.vue';
import Icon from 'dashboard/components-next/icon/Icon.vue';

const props = defineProps({
  label: {
    type: [String, Number],
    default: '',
  },
  variant: {
    type: String,
    default: 'solid',
    validator: value =>
      ['solid', 'outline', 'faded', 'link', 'ghost'].includes(value),
  },
  color: {
    type: String,
    default: 'blue',
    validator: value =>
      ['blue', 'ruby', 'amber', 'slate', 'teal'].includes(value),
  },
  size: {
    type: String,
    default: 'md',
    validator: value => ['xs', 'sm', 'md', 'lg'].includes(value),
  },
  icon: {
    type: String,
    default: '',
  },
  trailingIcon: {
    type: Boolean,
    default: false,
  },
  isLoading: {
    type: Boolean,
    default: false,
  },
});

const slots = useSlots();

const STYLE_CONFIG = {
  colors: {
    blue: {
      solid: 'bg-n-brand text-white hover:brightness-110 outline-transparent',
      faded:
        'bg-n-brand/10 text-n-blue-text hover:bg-n-brand/20 outline-transparent',
      outline: 'text-n-blue-text outline-n-blue-border',
      link: 'text-n-blue-text hover:underline outline-transparent',
    },
    ruby: {
      solid: 'bg-n-ruby-9 text-white hover:bg-n-ruby-10 outline-transparent',
      faded:
        'bg-n-ruby-9/10 text-n-ruby-11 hover:bg-n-ruby-9/20 outline-transparent',
      outline: 'text-n-ruby-11 hover:bg-n-ruby-9/10 outline-n-ruby-8',
      link: 'text-n-ruby-9 hover:underline outline-transparent',
    },
    amber: {
      solid: 'bg-n-amber-9 text-white hover:bg-n-amber-10 outline-transparent',
      faded:
        'bg-n-amber-9/10 text-n-slate-12 hover:bg-n-amber-9/20 outline-transparent',
      outline: 'text-n-amber-11 hover:bg-n-amber-9/10 outline-n-amber-9',
      link: 'text-n-amber-9 hover:underline outline-transparent',
    },
    slate: {
      solid:
        'bg-n-solid-3 dark:hover:bg-n-solid-2 hover:bg-n-alpha-2 text-n-slate-12 outline-n-container',
      faded:
        'bg-n-slate-9/10 text-n-slate-12 hover:bg-n-slate-9/20 outline-transparent',
      outline: 'text-n-slate-11 outline-n-strong hover:bg-n-slate-9/10',
      link: 'text-n-slate-11 hover:text-n-slate-12 hover:underline outline-transparent',
    },
    teal: {
      solid: 'bg-n-teal-9 text-white hover:bg-n-teal-10 outline-transparent',
      faded:
        'bg-n-teal-9/10 text-n-slate-12 hover:bg-n-teal-9/20 outline-transparent',
      outline: 'text-n-teal-11 hover:bg-n-teal-9/10 outline-n-teal-9',
      link: 'text-n-teal-9 hover:underline outline-transparent',
    },
  },
  sizes: {
    regular: {
      xs: 'h-6 px-2',
      sm: 'h-8 px-3',
      md: 'h-10 px-4',
      lg: 'h-12 px-5',
    },
    iconOnly: {
      xs: 'h-6 w-6 p-0',
      sm: 'h-8 w-8 p-0',
      md: 'h-10 w-10 p-0',
      lg: 'h-12 w-12 p-0',
    },
    link: {
      xs: 'p-0',
      sm: 'p-0',
      md: 'p-0',
      lg: 'p-0',
    },
  },
  fontSize: {
    xs: 'text-xs',
    sm: 'text-sm',
    md: 'text-sm font-medium',
    lg: 'text-base',
  },
  base: 'inline-flex items-center justify-center min-w-0 gap-2 transition-all duration-200 ease-in-out border-0 rounded-lg outline-1 outline disabled:cursor-not-allowed disabled:pointer-events-none disabled:opacity-50',
};

const variantClasses = computed(() => {
  const variantMap = {
    ghost: 'text-n-slate-12 hover:bg-n-alpha-2 outline-transparent',
    link: `${STYLE_CONFIG.colors[props.color].link} p-0 font-medium underline-offset-4`,
    outline: STYLE_CONFIG.colors[props.color].outline,
    faded: STYLE_CONFIG.colors[props.color].faded,
    solid: STYLE_CONFIG.colors[props.color].solid,
  };

  return variantMap[props.variant];
});

const isIconOnly = computed(() => !props.label && !slots.default);
const isLink = computed(() => props.variant === 'link');

const buttonClasses = computed(() => {
  const sizeConfig = isIconOnly.value ? 'iconOnly' : 'regular';
  const classes = [
    variantClasses.value,
    props.variant !== 'link' && STYLE_CONFIG.sizes[sizeConfig][props.size],
  ].filter(Boolean);

  return classes.join(' ');
});

const linkButtonClasses = computed(() => {
  const classes = [
    variantClasses.value,
    STYLE_CONFIG.sizes.link[props.size],
  ].filter(Boolean);

  return classes.join(' ');
});
</script>

<template>
  <button
    :class="{
      [STYLE_CONFIG.base]: true,
      [isLink ? linkButtonClasses : buttonClasses]: true,
      [STYLE_CONFIG.fontSize[size]]: true,
      'flex-row-reverse': trailingIcon && !isIconOnly,
    }"
  >
    <slot v-if="(icon || $slots.icon) && !isLoading" name="icon">
      <Icon :icon="icon" class="flex-shrink-0" />
    </slot>

    <Spinner v-if="isLoading" class="!w-5 !h-5 flex-shrink-0" />

    <slot v-if="label || $slots.default" name="default">
      <span v-if="label" class="min-w-0 truncate">{{ label }}</span>
    </slot>
  </button>
</template>
