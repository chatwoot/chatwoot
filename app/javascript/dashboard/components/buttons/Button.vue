<template>
  <button :type="type" class="button nice" :class="variant" @click="onClick">
    <fluent-icon
      v-if="!isLoading && icon"
      class="icon"
      :class="buttonIconClass"
      :icon="icon"
    />
    <spinner v-if="isLoading" />
    <slot />
  </button>
</template>

<script>
export default {
  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
    buttonIconClass: {
      type: String,
      default: '',
    },
    type: {
      type: String,
      default: 'button',
    },
    variant: {
      type: String,
      default: 'primary',
    },
  },
  methods: {
    onClick(e) {
      this.$emit('click', e);
    },
  },
};
</script>
