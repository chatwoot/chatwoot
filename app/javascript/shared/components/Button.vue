<template>
  <button
    :class="buttonClassName"
    :style="buttonStyles"
    :disabled="disabled"
    @click="onClick"
  >
    <slot />
  </button>
</template>
<script>
export default {
  props: {
    block: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String,
      default: 'blue',
    },
    bgColor: {
      type: String,
      default: '',
    },
    textColor: {
      type: String,
      default: '',
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    buttonClassName() {
      let className = 'text-white py-3 px-4 rounded shadow-sm';
      if (this.type === 'clear') {
        className = 'flex mx-auto mt-4 text-xs w-auto text-black-600';
      }

      if (this.type === 'blue' && !Object.keys(this.buttonStyles).length) {
        className = `${className} bg-woot-500 hover:bg-woot-700`;
      }
      if (this.block) {
        className = `${className} w-full`;
      }
      return className;
    },
    buttonStyles() {
      const styles = {};
      if (this.bgColor) {
        styles.backgroundColor = this.bgColor;
      }
      if (this.textColor) {
        styles.color = this.textColor;
      }
      return styles;
    },
  },
  methods: {
    onClick(e) {
      this.$emit('click', e);
    },
  },
};
</script>
