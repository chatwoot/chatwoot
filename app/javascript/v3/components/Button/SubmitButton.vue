<template>
  <button
    :type="type"
    data-testid="submit_button"
    :disabled="disabled"
    :class="computedClass"
    class="flex items-center w-full justify-center rounded-md bg-woot-500 py-3 px-3 text-base font-medium text-white shadow-sm hover:bg-woot-600 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-woot-500 cursor-pointer"
    @click="onClick"
  >
    <span>{{ buttonText }}</span>
    <fluent-icon v-if="!!iconClass" :icon="iconClass" class="icon" />
    <spinner v-if="loading" />
  </button>
</template>

<script>
import Spinner from 'shared/components/Spinner.vue';

export default {
  components: {
    Spinner,
  },
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    loading: {
      type: Boolean,
      default: false,
    },
    buttonText: {
      type: String,
      default: '',
    },
    buttonClass: {
      type: String,
      default: '',
    },
    iconClass: {
      type: String,
      default: '',
    },
    type: {
      type: String,
      default: 'submit',
    },
  },
  computed: {
    computedClass() {
      return `
        ${this.disabled ? 'opacity-40 hover:bg-woot-500' : ''}
        ${this.buttonClass || ' '}
      `;
    },
  },
  methods: {
    onClick() {
      this.$emit('click');
    },
  },
};
</script>
