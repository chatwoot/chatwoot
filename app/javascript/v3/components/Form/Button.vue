<template>
  <button
    class="inline-flex items-center gap-1 text-sm font-medium reset-base rounded-xl w-fit"
    :class="buttonClasses"
    v-bind="$attrs"
    @click="onClick"
  >
    <fluent-icon
      v-if="icon && !trailingIcon"
      size="1.16em"
      :icon="icon"
      class="flex-shrink-0"
    />
    <span
      v-if="$slots.default"
      class="text-sm font-medium truncate ltr:text-left rtl:text-right"
    >
      <slot />
    </span>
    <fluent-icon
      v-if="icon && trailingIcon"
      size="1.16em"
      :icon="icon"
      class="flex-shrink-0"
    />
  </button>
</template>
<script setup>
import { computed, useAttrs } from 'vue';

const props = defineProps({
  variant: {
    type: String,
    default: 'solid',
    validator: value => ['outline', 'ghost', 'solid'].includes(value),
  },
  size: {
    type: String,
    default: 'large',
    validator: value => ['medium', 'large'].includes(value),
  },
  icon: {
    type: String,
    default: '',
  },
  colorScheme: {
    type: String,
    default: 'primary',
    validator: value => ['primary', 'secondary', 'danger'].includes(value),
  },
  trailingIcon: {
    type: Boolean,
    default: false,
  },
});

const attrs = useAttrs();

const baseClasses = {
  outline: 'outline outline-1 -outline-offset-1',
  ghost: 'hover:text-600 active:text-600 focus:outline focus:outline-offset-1',
  solid:
    'hover:bg-700 active:bg-700 focus:outline focus:outline-offset-1 focus:outline-2',
};

const colorClass = computed(() => {
  if (attrs.disabled) {
    return 'bg-ash-200 text-ash-600 cursor-not-allowed';
  }

  const styleMap = {
    primary: {
      outline: `${baseClasses.outline} outline-primary-400 hover:text-primary-600 active:text-primary-600`,
      ghost: `${baseClasses.ghost} focus:outline-primary-400`,
      solid: `${baseClasses.solid} bg-primary-600 text-white focus:outline-primary-400`,
    },
    secondary: {
      outline: `${baseClasses.outline} outline-ash-400 hover:text-ash-900 active:text-ash-900 `,
      ghost: `${baseClasses.ghost} focus:outline-ash-400`,
      solid: `${baseClasses.solid} bg-ash-100 text-ash-900 focus:outline-ash-400`,
    },
    danger: {
      outline: `${baseClasses.outline} outline-ruby-400 hover:text-ruby-600 active:text-ruby-600`,
      ghost: `${baseClasses.ghost} focus:outline-ruby-400`,
      solid: `${baseClasses.solid} bg-ruby-600 text-white focus:outline-ruby-400`,
    },
  };

  const schemeStyles = styleMap[props.colorScheme];
  return schemeStyles[props.variant] || schemeStyles.solid;
});

const sizeClass = computed(() => {
  if (props.size === 'medium') {
    return 'h-8 px-3 py-1.5 text-sm';
  }
  return 'h-10 px-4 py-2.5 text';
});

const buttonClasses = computed(() => [colorClass.value, sizeClass.value]);

const emit = defineEmits(['click']);

const onClick = () => {
  if (props.disabled) {
    return;
  }
  emit('click');
};
</script>
