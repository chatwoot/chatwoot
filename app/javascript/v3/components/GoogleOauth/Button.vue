<template>
  <div class="flex flex-col">
    <a
      :href="getGoogleAuthUrl()"
      class="inline-flex w-full justify-center rounded-md bg-white py-3 px-4 shadow-sm ring-1 ring-inset ring-slate-200 dark:ring-slate-600 hover:bg-slate-50 focus:outline-offset-0 dark:bg-slate-700 dark:hover:bg-slate-700"
    >
      <img src="/assets/images/auth/google.svg" alt="Google Logo" class="h-6" />
      <span class="text-base font-medium ml-2 text-slate-600 dark:text-white">
        {{ $t('LOGIN.OAUTH.GOOGLE_LOGIN') }}
      </span>
    </a>
    <simple-divider
      v-if="showSeparator"
      ref="divider"
      :label="$t('COMMON.OR')"
      class="uppercase"
    />
  </div>
</template>

<script>
import SimpleDivider from '../Divider/SimpleDivider.vue';

export default {
  components: {
    SimpleDivider,
  },
  props: {
    showSeparator: {
      type: Boolean,
      default: true,
    },
  },
  methods: {
    getGoogleAuthUrl() {
      // Ideally a request to /auth/google_oauth2 should be made
      // Creating the URL manually because the devise-token-auth with
      // omniauth has a standing issue on redirecting the post request
      // https://github.com/lynndylanhurley/devise_token_auth/issues/1466
      const baseUrl =
        'https://accounts.google.com/o/oauth2/auth/oauthchooseaccount';
      const clientId = window.chatwootConfig.googleOAuthClientId;
      const redirectUri = window.chatwootConfig.googleOAuthCallbackUrl;
      const responseType = 'code';
      const scope = 'email profile';

      // Build the query string
      const queryString = new URLSearchParams({
        client_id: clientId,
        redirect_uri: redirectUri,
        response_type: responseType,
        scope: scope,
      }).toString();

      // Construct the full URL
      return `${baseUrl}?${queryString}`;
    },
  },
};
</script>
