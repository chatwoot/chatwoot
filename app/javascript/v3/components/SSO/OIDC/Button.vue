<template>
    <div class="flex flex-col">
        <simple-divider v-if="showSeparator" ref="divider" :label="$t('COMMON.OR')" class="uppercase" />
        <a :href="oidcLogin()"
            class="inline-flex w-full justify-center rounded-md bg-white py-3 px-4 shadow-sm ring-1 ring-inset ring-slate-200 dark:ring-slate-600 hover:bg-slate-50 focus:outline-offset-0 dark:bg-slate-700 dark:hover:bg-slate-700">
            <img src="/assets/images/auth/oidc.svg" alt="OIDC Logo" class="h-6" />
            <span class="text-base font-medium ml-2 text-slate-600 dark:text-white">
                {{ $t('LOGIN.OIDC.OIDC_LOGIN') }}
            </span>
        </a>
    </div>
</template>
  
<script>
import SimpleDivider from '../../Divider/SimpleDivider.vue';

export default {
    components: {
        SimpleDivider,
    },
    props: {
        showSeparator: {
            type: Boolean,
            default: true,
        },
    },
    methods: {
        oidcLogin() {
            const clientId= window.chatwootConfig.keycloakClientId;
            const redirectUri =  window.chatwootConfig.keycloakCallbackUrl

            const baseUrl = 'https://sso.onehash.ai/realms/OneHash/protocol/openid-connect/auth';
            const responseType = 'code';
            const scope = 'openid';

            const queryString = new URLSearchParams({
                client_id: clientId,
                redirect_uri: redirectUri,
                response_type: responseType,
                scope: scope,
            }).toString();

            return `${baseUrl}?${queryString}`;
        },
    },
};
</script>
  