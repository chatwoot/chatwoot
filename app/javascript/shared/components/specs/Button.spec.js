import { mount } from '@vue/test-utils';
import Button from '../Button.vue';

describe('Button', () => {
  let wrapper;

  const createComponent = (props = {}) => {
    wrapper = mount(Button, {
      props,
      slots: {
        default: 'Click me',
      },
    });
  };

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
    }
  });

  describe('rendering', () => {
    it('renders button with default props', () => {
      createComponent();

      expect(wrapper.find('button').exists()).toBe(true);
      expect(wrapper.text()).toBe('Click me');
      expect(wrapper.classes()).toContain('bg-woot-500');
      expect(wrapper.classes()).toContain('hover:bg-woot-700');
    });

    it('renders button with custom slot content', () => {
      wrapper = mount(Button, {
        slots: {
          default: '<span>Custom Content</span>',
        },
      });

      expect(wrapper.html()).toContain('<span>Custom Content</span>');
    });

    it('matches snapshot with default props', () => {
      createComponent();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('props', () => {
    it('applies block class when block prop is true', () => {
      createComponent({ block: true });

      expect(wrapper.classes()).toContain('w-full');
    });

    it('does not apply block class when block prop is false', () => {
      createComponent({ block: false });

      expect(wrapper.classes()).not.toContain('w-full');
    });

    it('applies clear type styling', () => {
      createComponent({ type: 'clear' });

      expect(wrapper.classes()).toContain('text-black-600');
      expect(wrapper.classes()).toContain('text-xs');
      expect(wrapper.classes()).not.toContain('bg-woot-500');
    });

    it('applies blue type styling by default', () => {
      createComponent({ type: 'blue' });

      expect(wrapper.classes()).toContain('bg-woot-500');
      expect(wrapper.classes()).toContain('hover:bg-woot-700');
    });

    it('applies custom background color when bgColor prop is provided', () => {
      createComponent({ bgColor: '#ff0000' });

      expect(wrapper.attributes('style')).toContain(
        'background-color: rgb(255, 0, 0)'
      );
    });

    it('applies custom text color when textColor prop is provided', () => {
      createComponent({ textColor: '#00ff00' });

      expect(wrapper.attributes('style')).toContain('color: rgb(0, 255, 0)');
    });

    it('applies both custom colors when both props are provided', () => {
      createComponent({
        bgColor: '#ff0000',
        textColor: '#00ff00',
      });

      const style = wrapper.attributes('style');
      expect(style).toContain('background-color: rgb(255, 0, 0)');
      expect(style).toContain('color: rgb(0, 255, 0)');
    });

    it('sets disabled attribute when disabled prop is true', () => {
      createComponent({ disabled: true });

      expect(wrapper.attributes('disabled')).toBeDefined();
      expect(wrapper.classes()).toContain('disabled:opacity-50');
    });

    it('does not set disabled attribute when disabled prop is false', () => {
      createComponent({ disabled: false });

      expect(wrapper.attributes('disabled')).toBeUndefined();
    });
  });

  describe('computed properties', () => {
    it('computes buttonClassName correctly for default blue type', () => {
      createComponent({ type: 'blue' });

      const expectedClasses = [
        'text-white',
        'py-3',
        'px-4',
        'rounded-lg',
        'shadow-sm',
        'leading-4',
        'cursor-pointer',
        'disabled:opacity-50',
        'bg-woot-500',
        'hover:bg-woot-700',
      ];

      expectedClasses.forEach(className => {
        expect(wrapper.classes()).toContain(className);
      });
    });

    it('computes buttonClassName correctly for clear type', () => {
      createComponent({ type: 'clear' });

      const expectedClasses = [
        'flex',
        'mx-auto',
        'mt-4',
        'text-xs',
        'leading-3',
        'w-auto',
        'text-black-600',
      ];

      expectedClasses.forEach(className => {
        expect(wrapper.classes()).toContain(className);
      });
    });

    it('computes buttonStyles correctly when custom colors are provided', () => {
      createComponent({
        bgColor: '#123456',
        textColor: '#abcdef',
      });

      const computedStyles = wrapper.vm.buttonStyles;
      expect(computedStyles.backgroundColor).toBe('#123456');
      expect(computedStyles.color).toBe('#abcdef');
    });

    it('returns empty buttonStyles when no custom colors are provided', () => {
      createComponent();

      const computedStyles = wrapper.vm.buttonStyles;
      expect(Object.keys(computedStyles)).toHaveLength(0);
    });
  });

  describe('accessibility', () => {
    it('maintains button semantics', () => {
      createComponent();

      expect(wrapper.element.tagName).toBe('BUTTON');
    });

    it('supports disabled state for screen readers', () => {
      createComponent({ disabled: true });

      expect(wrapper.attributes('disabled')).toBeDefined();
    });
  });

  describe('edge cases', () => {
    it('handles empty slot content gracefully', () => {
      wrapper = mount(Button, {
        slots: {
          default: '',
        },
      });

      expect(wrapper.text()).toBe('');
      expect(wrapper.find('button').exists()).toBe(true);
    });

    it('handles undefined props gracefully', () => {
      createComponent({
        bgColor: undefined,
        textColor: undefined,
        type: undefined,
      });

      expect(wrapper.find('button').exists()).toBe(true);
      // Should fall back to default blue styling
      expect(wrapper.classes()).toContain('bg-woot-500');
    });

    it('handles custom type that is not clear or blue', () => {
      createComponent({ type: 'custom' });

      // Should not apply blue styling when custom colors are not provided
      expect(wrapper.classes()).toContain('text-white');
      expect(wrapper.classes()).not.toContain('bg-woot-500');
    });
  });
});
