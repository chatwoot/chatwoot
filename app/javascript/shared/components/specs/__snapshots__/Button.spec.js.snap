// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > rendering > matches snapshot with default props 1`] = `
<button
  class="text-white py-3 px-4 rounded-lg shadow-sm leading-4 cursor-pointer disabled:opacity-50 bg-woot-500 hover:bg-woot-700"
>
  
  Click me
  
</button>
`;
